File: go/passport/passport_test.go
package passport

import (
	"testing"

	"github.com/cezexPL/agent-passport-standard/go/crypto"
)

func testConfig() Config {
	return Config{
		ID:        "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
		PublicKey: "z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
		OwnerDID:  "did:key:z6MkpOwnerABCDEFGHIJKLMNOPQRSTUVWXYZ12345",
		Skills: []Skill{
			{
				Name:         "go-developer",
				Version:      "1.0.0",
				Description:  "Go backend development",
				Capabilities: []string{"code_write", "test_run"},
				Hash:         "0x0000000000000000000000000000000000000000000000000000000000000001",
			},
		},
		Soul: Soul{
			Personality: "focused",
			WorkStyle:   "test-first",
			Constraints: []string{"deterministic"},
			Hash:        "0x0000000000000000000000000000000000000000000000000000000000000002",
			Frozen:      false,
		},
		Policies: Policies{
			PolicySetHash: "0x0000000000000000000000000000000000000000000000000000000000000003",
			Summary:       []string{"can_bid"},
		},
		Lineage: Lineage{
			Kind:       "single",
			Parents:    []string{},
			Generation: 0,
		},
	}
}

func TestNew(t *testing.T) {
	p, err := New(testConfig())
	if err != nil {
		t.Fatal(err)
	}
	if p.Type != "AgentPassport" {
		t.Errorf("type = %s, want AgentPassport", p.Type)
	}
	if p.Snapshot.Version != 1 {
		t.Errorf("version = %d, want 1", p.Snapshot.Version)
	}
	if p.Snapshot.Hash == "" {
		t.Error("snapshot hash should not be empty")
	}
	if p.GenesisOwner.Immutable != true {
		t.Error("genesis_owner.immutable should be true")
	}
}

func TestNew_MissingID(t *testing.T) {
	cfg := testConfig()
	cfg.ID = ""
	_, err := New(cfg)
	if err == nil {
		t.Error("should fail without id")
	}
}

func TestHash_Deterministic(t *testing.T) {
	p, _ := New(testConfig())
	h1, err := p.Hash()
	if err != nil {
		t.Fatal(err)
	}
	h2, err := p.Hash()
	if err != nil {
		t.Fatal(err)
	}
	if h1 != h2 {
		t.Errorf("non-deterministic: %s != %s", h1, h2)
	}
}

func TestSignVerify(t *testing.T) {
	pub, priv, err := crypto.GenerateKeyPair()
	if err != nil {
		t.Fatal(err)
	}
	p, err := New(testConfig())
	if err != nil {
		t.Fatal(err)
	}
	if err := p.Sign(priv); err != nil {
		t.Fatal(err)
	}
	if p.Proof == nil {
		t.Fatal("proof should be set")
	}
	ok, err := p.Verify(pub)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Error("signature should be valid")
	}
}

func TestSignVerify_WrongKey(t *testing.T) {
	_, priv, _ := crypto.GenerateKeyPair()
	pub2, _, _ := crypto.GenerateKeyPair()

	p, _ := New(testConfig())
	p.Sign(priv)

	ok, err := p.Verify(pub2)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Error("should not verify with wrong key")
	}
}

func TestFromJSON_Roundtrip(t *testing.T) {
	p, _ := New(testConfig())
	data, err := p.JSON()
	if err != nil {
		t.Fatal(err)
	}
	p2, err := FromJSON(data, false)
	if err != nil {
		t.Fatal(err)
	}
	if p2.ID != p.ID {
		t.Errorf("id mismatch: %s != %s", p2.ID, p.ID)
	}
	if p2.Snapshot.Hash != p.Snapshot.Hash {
		t.Error("snapshot hash mismatch after roundtrip")
	}
}

File: typescript/tests/passport.test.ts
import { describe, it, expect } from 'vitest';
import { AgentPassport } from '../src/passport.js';
import { generateKeyPair, snapshotHash } from '../src/crypto.js';
import type { PassportConfig } from '../src/types.js';

function makeConfig(publicKey: string): PassportConfig {
  return {
    id: 'did:key:z6MkTest',
    publicKey,
    ownerDID: 'did:key:z6MkOwner',
    skills: [
      { name: 'go-developer', version: '1.0.0', description: 'Go backend', capabilities: ['code_write'], hash: '0xabc' },
    ],
    soul: { personality: 'focused', work_style: 'test-first', constraints: [], hash: '0xdef', frozen: false },
    policies: { policy_set_hash: '0x123', summary: ['can_bid'] },
    lineage: { kind: 'original', parents: [], generation: 0 },
  };
}

describe('AgentPassport', () => {
  it('create → sign → verify → hash matches', async () => {
    const { publicKey, privateKey } = generateKeyPair();
    const { bytesToHex } = await import('@noble/hashes/utils');
    const cfg = makeConfig(bytesToHex(publicKey));

    const passport = await AgentPassport.create(cfg);
    expect(passport.data.snapshot.version).toBe(1);
    expect(passport.data.snapshot.hash).toBeTruthy();

    const hashBefore = passport.hash();
    await passport.sign(privateKey);
    expect(passport.data.proof).toBeTruthy();

    const valid = await passport.verify(publicKey);
    expect(valid).toBe(true);

    // Hash should be consistent (excluding proof)
    const hashAfter = passport.hash();
    expect(hashAfter).toBe(hashBefore);
  });

  it('frozen skills enforcement', async () => {
    const { publicKey } = generateKeyPair();
    const { bytesToHex } = await import('@noble/hashes/utils');
    const cfg = makeConfig(bytesToHex(publicKey));

    const passport = await AgentPassport.create(cfg);
    passport.data.snapshot.skills.frozen = true;

    expect(() =>
      passport.addSkill({ name: 'new-skill', version: '1.0.0', description: 'test', capabilities: [], hash: '0x1' })
    ).toThrow('skills are frozen');
  });

  it('snapshot hash chain (version 1→2)', async () => {
    const { publicKey } = generateKeyPair();
    const { bytesToHex } = await import('@noble/hashes/utils');
    const cfg = makeConfig(bytesToHex(publicKey));

    const passport = await AgentPassport.create(cfg);
    const v1Hash = passport.data.snapshot.hash;
    expect(passport.data.snapshot.prev_hash).toBeNull();

    passport.newSnapshot();
    expect(passport.data.snapshot.version).toBe(2);
    expect(passport.data.snapshot.prev_hash).toBe(v1Hash);
  });

  it('fromJson/toJson roundtrip', async () => {
    const { publicKey } = generateKeyPair();
    const { bytesToHex } = await import('@noble/hashes/utils');
    const cfg = makeConfig(bytesToHex(publicKey));

    const passport = await AgentPassport.create(cfg);
    const json = passport.toJson();
    const restored = AgentPassport.fromJson(json);
    expect(restored.data.id).toBe(passport.data.id);
    expect(restored.hash()).toBe(passport.hash());
  });
});
